### PR TITLE
Refactor command execution and ensure cleanup of the process group [ONPREM-1263]

### PR DESCRIPTION
**Pull request by christian-stephen**
:gear: **Issue**

---

:gear: **Change Description**

**Acceptance Criteria**:

---

:white_check_mark: **Solution**

---

:question: **Testing**

- [ ] Created and updated tests where applicable

---

:book: **Documentation Updates**

- [ ] Updated related documentation, if applicable
- [ ] Updated [changelog](https://github.com/circleci/runner-init/blob/main/CHANGELOG.md)

### CODE REVIEW

File: task/cmd/cmd.go
Comment (christian-stephen) [inline task/cmd/cmd.go:1]: Moved most of the command setup and execution to its own package to help simplify the core orchestrator code.

File: task/orchestrator_test.go [hunk 1]
```diff
@@ -84,6 +87,30 @@ func TestOrchestrator(t *testing.T) {
 			gracePeriod: 2 * time.Second,
 			wantError:   "",
 		},
+		{
+			name:   "child processes are cleaned up",
```
Comment (christian-stephen) [inline task/orchestrator_test.go:91]: Added a test to ensure the child process group is indeed terminated on shutdown of the agent.

File: task/orchestrator_test.go [hunk 2]
```diff
@@ -30,10 +33,15 @@ func TestOrchestrator(t *testing.T) {
...
-		// Set the process reap time to 0 to prevent interference between test cases
-		reapTime = 0
+		// Reduce the process reap time to prevent interference between test cases
+		reapTime = 500 * time.Millisecond
 	})
 
+	// Re-parent any child processes to us to simulate the orchestrator being init
+	const PrSetChildSubreaper = 36
+	_, _, err := syscall.Syscall(syscall.SYS_PRCTL, PrSetChildSubreaper, uintptr(1), 0)
+	assert.Check(t, err == 0)
```
Comment (christian-stephen) [inline task/orchestrator_test.go:43]: Discovering this trick took a while, but it let's us ensure the reaper is actually working.

File: task/cmd/cmd.go
Comment (christian-stephen) [inline task/cmd/cmd.go:1]: Moved most of the command setup and execution to its own package to help simplify the core orchestrator code.

File: task/orchestrator_test.go [hunk 3]
```diff
@@ -84,6 +92,30 @@ func TestOrchestrator(t *testing.T) {
 			gracePeriod: 2 * time.Second,
 			wantError:   "",
 		},
+		{
+			name:   "zombie processes are reaped",
```
Comment (christian-stephen) [inline task/orchestrator_test.go:96]: Added a test to ensure zombie process are indeed reaped on shutdown of the agent.

File: task/orchestrator.go [hunk 4]
```diff
@@ -249,7 +163,7 @@ func (o *Orchestrator) handleRunErrors(ctx context.Context, err error) error {
 	return errors.Join(failErr, unclaimErr, err)
 }
 
-var reapTime = 10 * time.Second // can be overridden by tests
+var reapTime = 30 * time.Second // can be overridden by tests
```
Comment (christian-stephen) [inline task/orchestrator.go:166]: Extended this more to match the default grace period. It's non-blocking, anyway.

File: task/orchestrator_test.go [hunk 5]
```diff
@@ -103,19 +135,19 @@ func TestOrchestrator(t *testing.T) {
 			},
 		},
 		{
-			name:   "error: task agent misbehaving",
+			name:   "error: task agent panicking",
```
Comment (christian-stephen) [inline task/orchestrator_test.go:138]: Figure this is better test. Same thing, really.

**Review by atulsingh0 — APPROVED** [review]

**Review by ryanwohara — APPROVED** [review]

**Pull request merged**